File: 2021_summer_algorithm_study_intermediate/using_python/1766.py
"""
input :
4 2
4 2
3 1

output :
3 1 4 2
"""
import sys
# from collections import deque
import heapq
sys.stdin = open("input.txt", "r")
input = sys.stdin.readline


if __name__ == "__main__":
    N, M = map(int, input().rstrip().split())
    graph = [[] for _ in range(N + 1)]
    indeg = [0 for _ in range(N + 1)]
    for _ in range(M):
        a, b = map(int, input().rstrip().split())
        graph[a].append(b)
        indeg[b] += 1

    min_heap = []
    #que = deque()
    for i in range(1, N + 1):
        if indeg[i] == 0:
            heapq.heappush(min_heap, i)
    while len(min_heap) != 0:
        cur = heapq.heappop(min_heap)
        print(cur, end=' ')
        for nxt in graph[cur]:
            indeg[nxt] -= 1
            if indeg[nxt] == 0:
                heapq.heappush(min_heap, nxt)

File: 2021_summer_algorithm_study_intermediate/using_cpp/1766.cpp
/*
input :
3 2
1 3
2 3

output :
1 2 3
*/

#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
#include <functional>
using namespace std;

// 전역 변수 선언
int N, M;
int a, b;
vector<int> graph[32005];
int indeg[32005];
priority_queue<int, vector<int>, greater<int>> min_heap;


int main(){
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    freopen("input.txt", "r", stdin);

    cin >> N >> M;
    while (M--){
        cin >> a >> b;
        graph[a].push_back(b);
        indeg[b] += 1;
    }

    for (int i = 1; i <= N; i++){
        if (indeg[i] == 0){
            min_heap.push(i);
        }
    }

    while (min_heap.size() != 0){
        int cur = min_heap.top(); min_heap.pop();
        cout << cur << ' ';
        for (int nxt: graph[cur]){
            indeg[nxt] -= 1;
            if (indeg[nxt] == 0){
                min_heap.push(nxt);
            }
        }
    }

    return 0;
}